proposal multi version (#110)
gardener/landscaper-service: docs/proposals/multi-version.md

Service (LaaS) landscape.

## Motivation

A LaaS landscape should support different versions of a Landscaper instance (LI), i.e. different LandscaperDeployments
install different versions of a [this](../../.landscaper/landscaper-instance) component. A LaaS landscape should support 
a predefined set of LI versions, which change from release to release (probably if the LaaS component version changes)
and a customer could select when to upgrade which of his instances.

A LI is a complex software artefact used by the customers for their deployments. Upgrading a LI might include complex
things like updating the used Helm version, the kubernetes version of the Landscaper resource shoot cluster and many
more. Due to this complexity, it is impossible to ensure that LI updates might not harm an existing customer deployments.

To reduce this risk, customers must have possibility to test a new LI component version with their development setup 
first, before they decide to upgrade the LI component of their productive environment.

## Requirements

This chapter defines the rules for supported LI versions in a LaaS landscape:

- Versioning:

>   - Patch: A new patch version of a LI should only include bugfixes and security fixes. It should not break a customer

Reviewer: ```suggestion
  - Patch: A new patch version of a LI should only include bugfixes and security fixes. It must not break a customer
```

A patch release that would break a customer LI should be considere a bug and must be avoided. I think we should use strong wording here.

Author: I have changed it that way

    setup.
  - Minor: A new minor version of a LI could contain additional features, minor updates of the most important components,
    like Helm or the Garden shoot cluster. Due to its complexity a minor update might break a customer setup. 

>   - Major: Groundbreaking changes. Might break a customer setup.

Reviewer: ```suggestion
  - Major: Groundbreaking changes. Customer setup is very likely to be broken.
```

Author: I have changed it that way

- Supported LI version of a LaaS landscape:
  - The list of supported versions of a LaaS landscape only contains one patch version (usually the latest) for every 

>     supported minor version. 

Reviewer: I think this sentence draws an incomplete picture. You are describing the supported versions later.
So we either need to add the full description on what version range is supported or remove this section and rely und the upgrade description.

Author: I'm not sure if I got your point but I have reworked that part with examples and some motivation. Could you check if this is ok now?

Reviewer: Strange, I dont see any changes since my review.
By incomplete I meant, it doesnt tell how many minor versions/major we are supporting at the same time. However, this is described later on in the document. This made me wonder if we need this sentence at all, when there is a complete detailed description.

Author: Sorry now I have pushed it. I answered when I was working on the changes and push them afterwards.

Author: and I pressed the rereview button to early.

- Basic upgrading rules
  - A LI instance could be upgraded to the currently supported patch level of its current minor version. 

>   - If a LI instance runs on a currently supported patch level or its current minor version is not supported anymore,

Reviewer: ```suggestion
  - If a LI instance current minor version is not supported anymore,
```

I don't get what you mean by a supported patch level.

Author: I also needed some time to get it again but I guess it is right. :-) 

I have rephrased it but not sure if this is understandable now. The goal of these two definitions is that a version of LI should be upgraded first to the supported patch level of its minor level and if this has happened of if this is not possible, because there is no supported patch level for the current minor version, then an upgrade to the next minor version is allowed.

    then: 

>     - it could be upgraded to the next supported minor version or 

Reviewer: ```suggestion
    - it must be upgraded to the next supported minor version or 
```

Author: These rules try to define what is allowed as upgrade. Therefore I change it to "is allowed to" instead of "could" at the different places.

>     - if there is no supported higher minor version it could be upgraded to the lowest supported minor version of the 

Reviewer: ```suggestion
    - if there is no supported higher minor version it must be upgraded to the lowest supported minor version of the 
```

Author: same as before

      next major version.

- An upgrade of a LaaS landscape updates the list of supported LI versions according to the following rules:
  - A supported minor version is set on deprecated, if there is a supported version with a higher major version or
    a supported version with the same major but a higher minor version.

>   - A minor version could only be removed if it was deprecated for more than 3 months.

Reviewer: ```suggestion
  - A supported minor version can only be removed if it was deprecated for more than 3 months.
```

Reviewer: Shall we specify the deprecation time in a dedicated sentence and replace it in this sentence with `A minor version can only be removed if the deprecation duration has ended.`?
When we decide to change the deprecation duration, we dont have to search and replace in the text.
What do you think?

Author: I have changed the wording and introduced deprecation duration.

  - It is allowed to replace a minor version with one with a higher patch level but not with a lower one. Such a change 

>     will not affect if the minor version was deprecated or not.

Reviewer: Sorry, I don't understand the second sentence.

Author: I rephrased it.

  - Every LI in a LaaS landscape is automatically upgraded to the currently supported patch version of its current
    minor version.
  - A customer could decide if his LI instance is upgraded automatically to higher minor versions or if this should be
    triggered manually.
  - Upgrades to the next major version must be triggered manually.
  - Customers are automatically informed about newly supported and deprecated LI component versions. 
  - Customers are automatically informed when a LI version will be removed from the supported version list.
  - If a customer LI instance is running on an unsupported minor version, it is automatically upgraded,
    according to upgrading rules described above. The customer is automatically informed about this.

- A customer must be able 
  - to select its intended LI version during onboarding. This must be optional. He gets the latest version
    if nothing is selected.
  - to see
    - the current version of his LI
    - the supported versions
    - the deprecated supported versions
    - the exact timeframes when deprecated LI versions are removed (-> forced update)
    - the upgrade path according to the rules above

## Questions:

> - If we release new minor versions too frequently, the customer must test and upgrade quite often.

Reviewer: According to the rules above, a minor version can only removed when deprecated for 3 months.
This means, a customer can never be forced to upgrade more often than 3 months or am I wrong?
With the deprectation rule it also meants, that we can't remove any intermediate minor version up to the latest supported minor version (or next major version) according to the above rules.
Because otherwise the customer has no uniterrupted upgrade path.

I think this rather means for us that we have to be more carefully and planned about new minor and major releases. Because it always renders one minor version deprecated.

Author: I'm still thinking about this. If we start with v0.1.0 and release a new minor version every week and the customer is currently on the first minor version v0.1.0, he needs to do his first upgrade after 3 months. But then he has to upgrade from v0.1.0 to v0.2.0, then from v0.2.0 to v0.3.0 until the latest supported minor version to get more than 3 months before he has to care about his next upgrade. Version v0.2.0 for example will be removed in one weeks.

## Potential architectural consequences

- The Central Landscaper should always be on the latest Landscaper version. The version referenced by the

>   LaaS project.

Reviewer: This means we can't have breaking changes in the LI blueprints, right? Or the landscaper has to provide a conversion of incompatible blueprints etc.

Author: I wanted to express only that the Landscaper version of the central landscaper is independent of the supported versions of the LI component. I changed the wording.

# Multi-Version-Landscaper-Instances

This document contains a proposal how to support multiple versions of Landscaper Instances in one Landscaper as a 
Service (LaaS) landscape.

## Requirements

- A LaaS landscape should support different versions of a Landscaper instance (LI), i.e. different LandscaperDeployments 
  install different versions of a [this](../../.landscaper/landscaper-instance) component. A more detailed
  selection of particular deployer versions seems to be too complicated and out of scope.

- A LaaS landscape supports a predefined set of LI versions, which change from release to realease (probably if the LaaS 
  component version changes).

- The Central Landscaper should always be on the latest Landscaper version. The version referenced by the 
  LaaS project.  

- A customer must be able 
  - to select its intended LI version during onboarding. This must be optional. He gets the latest version
    if nothing is selected.
  - to upgrade to its intended LI version. Therefore, an allowed upgrade path must be enforced.

>   - see which versions are currently supported, which are recommended and which are outdated and what are the possible

Reviewer 2: And a customer must be able to see the exact timeframes in which
- specific LI versions are supported
- specific LI versions are set to "deprecated"
- specific LI versions run out of support -> results in a forced-upgrade ?
-

Author: I have included these requirements

    upgrade paths.

- How to upgrade a customer LI instance if his current version is not supported anymore? This might become even more 
  critical if a security issue came up.
  - Proposal for an initial solution: 

>     - Not supported versions are not upgraded. Only the reconciliation of the corresponding Installation is retriggered

Reviewer 2: What does this mean exactly ? I would like to see that we never have unsupported LI versions running somewhere. Stakeholders must make sure that they upgrade to a supported LI version until their currently used LI version runs out of support. If they don't do this -> forced-upgrade ?

Author: lets discuss this later

Author: I have adapted this according to the result of our discussion: forced update for unsupported versions

      for token rotation issues, or we enhance the Installation by some automatic reconcile mechanisms.
    - Affected customers are informed via email. This mail is send automatically and if the customer does not upgrade,

>       we get informed to personally contact him.

Reviewer 2: I would not like to have any manual steps here like contacting stakeholders personally. This might scale initially, but will not in the future. Automatic emails is ok, but if stakeholders do not upgrade, shouldn't we just force-upgrade ? For sure this has to be mentioned super prominently in our LaaS documentation, so that stakeholders know of this force-upgrade procedure if they don't react.

Author: let discuss this later

Author: adapted aas discussed

- Do we need some kind of automatic upgrading, which could be selected ba a customer? Could it be selected that only

>   patch or minor versions are automatically upgraded?

Reviewer 2: Yes, definitely. We should have a mechanism for stakeholders to choose if they want automatic patch or even minor version LI upgrades (and be able to specify when these upgrades should happen, e.g. day & time. Major version upgrades should not be done automatically, but have to be a decision by the stakeholder.

Author: let discuss this in the meeting as the points above - no two channels for now - I loose control :-)

Author: adapted as discussed
